proxy.go

```diff
@@ -408,6 +413,37 @@ func requireBasicAuth(authenticator *auth.BasicAuth, handler func(http.ResponseW
 	})
 }
 
+func WithSuitableAuthentication(authenticator *auth.BasicAuth, handler func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !guestAccessAllowed {
+			//All requests require authentication
+			delegate := requireBasicAuth(authenticator, handler)
+			delegate(w, r)
+		} else {
+			_, _, basicAuthPresent := r.BasicAuth()
+			if !basicAuthPresent {
+				//Run the handler as unauthenticated user
+				confLock.RLock()
+				_, ok := quota[guestUserName]
+				confLock.RUnlock()
+				if !ok {
+					reply(w, errMsg("Guest access is unavailable."), http.StatusUnauthorized)
+				} else {
+					handler(w, r)
+				}
+			} else {
+				//Check that authentication is correct
+				userName := authenticator.CheckAuth(r)
```

Comment:
Good idea. I didn't copy/paste your suggestion as-is, because I want to  avoid the call to `r.BasicAuth()` when guests are not allowed. 
```
if !guestAccessAllowed {
    //All requests require authentication
    ...
} else if _, _, basicAuthPresent := r.BasicAuth(); !basicAuthPresent {
    //Run the handler as unauthenticated user
    ...
} else {
    //Run the handler using basic authentication
    ...
}
```